rbd: Address rbd mirroring TODO's

Madhu-1 opened the pull request:
This PR fixes the TODO's mentioned in #1896 

updates #1896 

Below is the list of TODO's

- [x]  move getVolumeFromID to controllerserver.go and reuse it wherever it's applicable.
- [x] create a new Replication controller struct for the replication operations.
- [x] check do we need to return abort after disabling the mirroring
- [ ] use string instead of int for returning a valid error message
- [x] check if the image is not found return a valid error message
- [x] check the image state and return its ready to use or not
- [ ] recheck https://github.com/ceph/ceph-csi/pull/1877#discussion_r589327546
- [x] add unit tests for helper functions
- [ ] add e2e testing

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Note: AFAIK E2E cannot be added as we need two ceph clusters for testing mirroring

Madhu-1 committed:
rbd: Add ReplicationServer struct for replication operations

added ReplicationServer struct for the replication related
operation it also embed the ControllerServer which
already implements the helper functions like locking/unlocking etc.

removed getVolumeFromID and cleanup functions for better
code readability and easy maintaince.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: add a check for image mirror disabling state

the rbd mirror state can be in enabled,disabled
or disabling state. If the mirroring is not disabled
yet and still in disabling state. we need to
check for it and return abort error message if
the mirroring is still getting disabled.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: make replication operations as rbdImage methods

added replication related operations as a method
of rbdImage as these methods can be easily used
when we introduce volumesnaphot mirroring operations.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: check for peer site status

Do resync if the image is in unknow or in error
state.

Check for the current image state for up+stopped
or up+replaying and also all peer site status
should be un up+stopped to confirm that resyncing
is done and image can be promoted and used.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: remove todo for image not found

Incase of resync the image will get deleted, gets
recreated and its a a time consuming operation.
It makes sense to return aborted error instead
of not found as we have omap data only the image
is missing in rbd pool.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -158,6 +158,41 @@ func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.
 	return rbdVol, nil
 }
 
+// getVolumeFromID gets the rbd image details from the volumeID.
```
please mention in the comment that when calling this function, a `defer cs.cleanup()` should be done as well

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -158,6 +158,41 @@ func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.
+	return rbdVol, cr, err
+}
+
+// cleanup performs below resource cleanup operations.
```
Add a reference back to `cs.getVolumeFromID()`, only after calling that, this function should be called.

nixpanic commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -59,6 +59,18 @@ const (
 
+// ReplicationServer struct of rbd CSI driver with supported methods of Replication
+// controller server spec.
+type ReplicationServer struct {
```
Yay, much cleaner! :partying_face:

nixpanic commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -190,10 +190,22 @@ func (rs *ReplicationServer) DisableVolumeReplication(ctx context.Context,
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		if mirroringInfo.State == librbd.MirrorImageDisabling {
+			return nil, status.Errorf(codes.Aborted, "%s is in disabling state", volumeID)
```
I assume the consumer of the API (kube-storage sidecar?) will do a backoff and check later?

Madhu-1 replied:
Yes as its an Aborted error message the operator will take care of doing backoff check

nixpanic commented on internal/rbd/mirror.go:
```diff
@@ -118,6 +118,9 @@ type imageMirrorStatus struct {
 	Description string `json:"description"`
 	LastUpdate  string `json:"last_update"`
+	PeerSites   []struct {
+		State string `json:"state"`
```
Does this only contain `State`, and no identification of the peer in a particular state? It might be useful to mention which peer is in which state?

Madhu-1 replied:
Added more fields like siteName, Description, and LastUpdate time.

nixpanic commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -368,8 +368,26 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
 	if state == upAndStopped || state == upAndReplaying {
+		// Make sure the peer site image state is up and stopped
 		ready = true
+		for _, s := range mirrorStatus.PeerSites {
```
`PeerSites` describes multiple sites, if there is peer in a incomplete state, you may want to log that peer so users can investigate what the cause could be

Madhu-1 replied:
Added logging now.

nixpanic commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -336,9 +336,8 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
-		// TODO: check if the image is not found return valid error message
 		// in case of Resync the image will get deleted and gets recreated and
-		// it takes time for this operation
+		// it takes time for this operation.
```
Should that be a `codes.Internal` error? Or, should it be `Aborted` so a retry can be done later?

Madhu-1 replied:
Aborted makes sense. Thanks for the suggestion.

nixpanic reviewed (changes requested):
Mostly nits and questions.

Please use different PRs for addressing different TODO's. A single PR with cleanup and enhancements is more work to review than small PRs that address a single topic. No need to split this PR up, but please do not add additional commits.

Madhu-1 commented:
> Mostly nits and questions.
> 
> Please use different PRs for addressing different TODO's. A single PR with cleanup and enhancements is more work to review than small PRs that address a single topic. No need to split this PR up, but please do not add additional commits.

will keep this in mind  :+1:

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -158,6 +158,44 @@ func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.
+	// validate the volume ID
+	cr, err := util.NewUserCredentials(secrets)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, err.Error())
```
I wonder if you need to return a gRPC error here. The callers of `cs.getVolumeID()` seem to take the error, inspect it and return their own gRPC version.

Madhu-1 replied:
Yes removed GRPC errors from this function

nixpanic replied:
It is still there on line 167?

nixpanic replied:
uh, line 173: https://github.com/ceph/ceph-csi/pull/1924/commits/f9fb9b6cd260e40cb54d0d6386bfb6a4b92c2604#diff-3792fcc7313c4d1e0b81c4d3ff7ac714b9d299130641673729196b2a4a1d6f2aR173

nixpanic commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -368,8 +368,26 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
 		ready = true
+		for _, s := range mirrorStatus.PeerSites {
+			if imageMirroringState(s.State) != upAndStopped {
+				util.DebugLog(ctx, "peer site name=%s mirroring state=%s, description=%s and lastUpdate=%s", s.SiteName, s.State, s.Description, s.LastUpdate)
```
I always think of `DebugLog` for developers, and `DefaultLog` or `InfoLog` and higher for users. Do you want to assist users with troubleshooting potential mirroring issues?

Madhu-1 replied:
Moved to InfoLog

nixpanic reviewed (commented):
More nits after an other careful look through the changes

nixpanic reviewed (changes requested):
LGTM, except for the error return at https://github.com/ceph/ceph-csi/commit/f9fb9b6cd260e40cb54d0d6386bfb6a4b92c2604#diff-3792fcc7313c4d1e0b81c4d3ff7ac714b9d299130641673729196b2a4a1d6f2aR173

Madhu-1 commented:
> LGTM, except for the error return at [f9fb9b6#diff-3792fcc7313c4d1e0b81c4d3ff7ac714b9d299130641673729196b2a4a1d6f2aR173](https://github.com/ceph/ceph-csi/commit/f9fb9b6cd260e40cb54d0d6386bfb6a4b92c2604#diff-3792fcc7313c4d1e0b81c4d3ff7ac714b9d299130641673729196b2a4a1d6f2aR173)

Fixed now

Madhu-1 commented:
/test all

Madhu-1 commented:
/retest all

humblec commented on internal/rbd/controllerserver.go:
```diff
@@ -158,6 +152,39 @@ func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.
+		return nil, cr, util.ErrOperationExists
+	}
+
+	rbdVol, err := genVolFromVolID(ctx, volumeID, cr, secrets)
```
@Madhu-1 why dont we unlock or cleanup from this function if we are done with decoding ?

Madhu-1 replied:
the lock needs to exist till the last part of the RPC calls. unlocking or cleanup from this function is not possible

humblec replied:
just to understand this better, the lock is taken inside getVolumeFromID() and unlock happens from the caller buildcreateVolumeResponse. so if the lock is required for the entire RPC call, why dont we take the lock from the caller itself ? @Madhu-1

humblec replied:
Lock and unlock from either caller or callee is what I am suggesting.

Madhu-1 replied:
This was done to avoid a lot of code duplication which was suggested as an enhancement in the last PR. if required i can change it back. Please let me know  your suggestion

humblec replied:
I am in favor of lock and unlock in one/same block scope. Lets check with @nixpanic too

Madhu-1 replied:
@nixpanic what do you think?

nixpanic replied:
I prefer to have locks taken and released in the same function. It should be an exception where this is not done.

Maybe introduce a functions like:
```golang
func (cs *ControllerServer) lockVolume(volumeID string) error {
    ...
}

func (cs *ControllerServer) unlockVolume(volumeID string) error {
    ...
}
```

This makes it really easy to re-use the functions everywhere, and set a 
```golang
defer cs.unlockVolume(volID)
```
after obtaining the lock.

I understand/guess that the `cleanup()` function was introduced to reduce the code in certain functions. However, the function moves away from being clear on when/what actions are needed, making maintenance and understanding of the code harder.

If reducing the complexity/size of a function is needed, it is better to introduce (small) helper functions that handle part of the functions work.

Madhu-1 replied:
already lock and unlock are smaller functions, introducing one more wrapper around it is not a good idea. I removed the commit which does this enhancement for now for clean code maintenance and reading.

humblec commented on internal/rbd/controllerserver.go:
```diff
@@ -1130,6 +1128,13 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
 		return nil, err
 	}
 
+	// lock out volumeID for clone and delete operation
```
why we are taking expandLock here for clone and delete ?

Madhu-1 replied:
This is not a new implementation, this is existing and the lock is moved below.

humblec commented on internal/rbd/controllerserver_test.go:
```diff
@@ -0,0 +1,89 @@
+)
+
+func TestControllerServercleanup(t *testing.T) {
+	type fields struct {
```
somehow I am not in favor of this struct name.. Can we have a better meaningful name please?

Madhu-1 replied:
This auto-generated code. can you please suggest what's in your mind?

humblec commented on internal/rbd/mirror.go:
```diff
@@ -118,27 +118,33 @@ type imageMirrorStatus struct {
 	State       string `json:"state"` // rbd image state
 	Description string `json:"description"`
 	LastUpdate  string `json:"last_update"`
+	PeerSites   []struct {
```
Can any of the field can be empty?

Madhu-1 replied:
AFAIK, No this cannot be empty as at least one peer site will be present in case of mirroring.

humblec commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -208,10 +208,22 @@ func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
 	case librbd.MirrorImageDisabled:
-		// image is already in disabled state
+	// image mirroring is still disabling
+	case librbd.MirrorImageDisabling:
```
Please move `Disabled` and `Disabling` cases above `Enabled`. It will avoid extra operations if image is in any of the disabled/disabling state.

humblec commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -265,12 +292,27 @@ func (cs *ControllerServer) PromoteVolume(ctx context.Context,
+			util.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, volumeID)
+			err = status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+		case errors.Is(err, ErrImageNotFound):
+			err = status.Errorf(codes.NotFound, "volume %s not found", volumeID)
```
just thinking, `volume not found` or `image not found` is better here?

Madhu-1 replied:
as we are considering image as the volume, IMO volume should be fine here

humblec replied:
all the mirror operations log ( rbd_mirror*,)pointing/referencing `image` and here its volume or mix (like in DisableVolumeReplication()) at times. in case of rbd, volume and image is pretty much same, so I was thinking the uniqueness is good

Madhu-1 replied:
as it's a volumeID which we got from the request, sending an error as the volume Not found. all the functions in mirror.go are referring to the image because after decoding the volumeID we got the image name as those are internal functions and we are logging it as an image. Do you still think the error message we send back to the RPC calls should say `image not found` instead of `volume not found` , if yes,i can do it.

humblec replied:
bit difficult while we interchange the reference, but its fine for now .

Madhu-1 committed:
rbd: early check for disabled,disabling in DisableVolumeReplication

added early check for disabling and disabled
image mirroring state in DisableVolumeReplication

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
>  Warning  Failed     24s                kubelet            Error: ImagePullBackOff
  Normal   Pulling    10s (x2 over 25s)  kubelet            Pulling image "docker.io/ceph/ceph:v14.2.12
 Warning  Failed     10s (x2 over 25s)  kubelet            Failed to pull image "docker.io/ceph/ceph:v14.2.12": rpc error: code = Unknown desc = Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
  Warning  Failed     10s (x2 over 25s)  kubelet            Error: ErrImagePull

Rook deployment failed.

Madhu-1 commented:
/test ci/centos/upgrade-tests-rbd

Madhu-1 commented:
/test ci/centos/upgrade-tests-cephfs

humblec commented:
@Madhu-1 apart from lock/unlock comment (https://github.com/ceph/ceph-csi/pull/1924#discussion_r604633336 ) this looks good  Thanks

Madhu-1 commented:
> @Madhu-1 apart from lock/unlock comment ([#1924 (comment)](https://github.com/ceph/ceph-csi/pull/1924#discussion_r604633336) ) this looks good Thanks

@humblec This is fixed now, PTAL

humblec commented:
> > @Madhu-1 apart from lock/unlock comment ([#1924 (comment)](https://github.com/ceph/ceph-csi/pull/1924#discussion_r604633336) ) this looks good Thanks
> 
> @humblec This is fixed now, PTAL

Sure.. thanks!

humblec commented:
> > > @Madhu-1 apart from lock/unlock comment ([#1924 (comment)](https://github.com/ceph/ceph-csi/pull/1924#discussion_r604633336) ) this looks good Thanks
> > 
> > 
> > @humblec This is fixed now, PTAL
> 
> Sure.. thanks!

LGTM. Approved.. :+1:

Madhu-1 commented:
@mergify rebase
